### PR TITLE
WIP authors: profile pages

### PR DESCRIPTION
**Pull request by jacenkow**
* Adds Angular.js app in authors module.

* Adds an API endpoint for collecting all publications by a given author.

* Adds new templates for an author brief and detailed views.

Signed-off-by: Grzegorz Jacenków <grzegorz.jacenkow@cern.ch>

### CODE REVIEW

**Comment by jacenkow** [comment]
@jmartinm

**Comment by jacenkow** [comment]
It's green. :+1:

**Comment by kaplun** [comment]
Can you post screenshots of the current state?

**Comment by jacenkow** [comment]
Authors

**Comment by jacenkow** [comment]
screen shot 2016-03-21 at 15 00 25

CC: @kaplun

**Comment by eamonnmag** [comment]
You should add labels to the axes so you can see which is which. Also, these axes should be coloured in the shade of blue related to the type of information it is showing, as shown in the attached image.
screen shot 2016-03-21 at 17 58 43
screen shot 2016-03-21 at 17 58 53

**Comment by jacenkow** [comment]
@eamonnmag cheers for the mockups
Yeah, this is what exactly I wanted to do (having just a printed design in B&W), however still warming up to D3.js :smile:

**Comment by kaplun** [comment]
Playing the role of @salmele here :smile: : for colorblind people, is there enough information for distinguishing between the 2 shades of blue and make the correct association with the corresponding axes?

**Comment by eamonnmag** [comment]
The light blue would need to be a bit lighter to cover all types of colour blindness. You can test using this easy to use tool - http://www.color-blindness.com/coblis-color-blindness-simulator/

File: inspirehep/config.py [hunk 1]
```diff
@@ -40,6 +40,8 @@ def _(x):
 # =====
 DEBUG_TB_INTERCEPT_REDIRECTS = False
 SERVER_NAME = "localhost:5000"
+ASSETS_DEBUG = True
```
Comment (jmartinm) [inline inspirehep/config.py]: This config should be in your personal `var/inspirehep-instance/inspirehep.cfg`

File: inspirehep/config.py [hunk 2]
```diff
@@ -40,6 +40,8 @@ def _(x):
...
+DEBUG = True
```
Comment (jmartinm) [inline inspirehep/config.py]: These config variables should be in your personal var/inspirehep-instance/inspirehep.cfg

File: inspirehep/modules/authors/bundles.py [hunk 3]
```diff
@@ -23,10 +23,33 @@
...
     "scss/authors/authors-update-form.scss",
     output='gen/inspire-author-update.%(version)s.css',
     depends='scss/forms/*.scss',
     filters="scss, cleancss"
 )
+
+profile_css = Bundle(
+    "scss/authors/authors-profile.scss",
+    output='gen/authors.%(version)s.css',
+    depends='scss/forms/*.scss',
```
Comment (jmartinm) [inline inspirehep/modules/authors/bundles.py]: I guess it would depend on the `authors` folder instead.

File: inspirehep/modules/authors/bundles.py [hunk 4]
```diff
@@ -23,10 +23,33 @@
...
+js = NpmBundle(
+    'js/authors/app.js',
+    'js/author_settings.js',
+    filters='requirejs',
+    depends=(
+        'js/authors/*.js'
+    ),
+    output="gen/authors.%(version)s.js",
+    npm={
+        "jquery": "~1.9.1",
```
Comment (jmartinm) [inline inspirehep/modules/authors/bundles.py]: No need for jquery and angular to be added here as they are already present in other `npm` sections in other bundles.

File: inspirehep/modules/authors/static/js/author_settings.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * In applying this licence, CERN does not waive the privileges and immunities
+ * granted to it by virtue of its status as an Intergovernmental Organization
+ * or submit itself to any jurisdiction.
+
+ */
```
Comment (jmartinm) [inline inspirehep/modules/authors/static/js/author_settings.js]: Should be enough to use the `settings.js` from `inspire-next` no need to add a new one.

File: inspirehep/modules/authors/static/js/build.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Invenio; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+({
```
Comment (jmartinm) [inline inspirehep/modules/authors/static/js/build.js]: Here we should also use the `build.js` already present. You will be overwriting it with this file as they are both in `static/js/build.js`

File: inspirehep/modules/theme/ext.py [hunk 7]
```diff
@@ -59,6 +59,7 @@ def init_app(self, app, assets=None, **kwargs):
 
         # Add Jinja2 extensions.
         app.jinja_env.add_extension('jinja2.ext.do')
+        app.jinja_env.autoescape = False
```
Comment (jmartinm) [inline inspirehep/modules/theme/ext.py]: Have a look if there is a better method for escaping for your needs instead of globally disabling http://flask.pocoo.org/docs/0.10/templating/#controlling-autoescaping

File: inspirehep/modules/theme/templates/inspirehep_theme/format/record/Base_HTML_brief.html [hunk 8]
```diff
@@ -0,0 +1,56 @@
+{#
```
Comment (jmartinm) [inline inspirehep/modules/theme/templates/inspirehep_theme/format/record/Base_HTML_brief.html]: Note that the brief formats are now using purely AngularJS for display so they can be removed from the PR.

**Comment by jmartinm** [comment]
Code looking good :+1:  Let me know if it is ready to be installed locally and give feedback.

File: inspirehep/modules/theme/ext.py (hunk 7 above)
Comment (kaplun) [inline inspirehep/modules/theme/ext.py]: Actually we should really keep autoescape to `True` in order to avoid XSS slipping through

**Comment by jacenkow** [comment]
profile

**Comment by eamonnmag** [comment]
@jacenkow - padding needs to be added to the research works so that the table is not flush with the bottom edge.

**Comment by jacenkow** [comment]
Blocker / missing data:
- [ ] co-authors
- [ ] citations
- [ ] HEP index
- [ ] DataTable (for listing publications)

To do:
- [ ] proper styling of statistics graph
- [ ] hookup publication type (journal, conference paper, etc.)
- [ ] refactor stylesheet

File: inspirehep/modules/authors/static/js/authors/publications.js [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+          year: parseInt(year),
+          publications: count
+        });
+      });      
+
+      return [publicationsDataToPlot, researchWorks];
+    };
+
+    var recid = $scope.record.control_number;
+    $http.get("/author/publications?recid=" + recid)
```
Comment (jacenkow) [inline inspirehep/modules/authors/static/js/authors/publications.js]: @jmartinm this one is called every time when there is a reference form a directive.

File: inspirehep/modules/authors/static/js/authors/author.js [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+  });
+
+  app.directive('authorEducation', function() {
+     return {
+      restrict: 'E',
+      templateUrl: '/static/js/authors/templates/position.html',
+      scope: true,
+      controller: 'AuthorController',
+      link: function(scope) {
+        scope.data = scope.positions.education;
```
Comment (jmartinm) [inline inspirehep/modules/authors/static/js/authors/author.js]: should it be a more specific attribute inside of `scope`? Instead of using `.data`

File: inspirehep/modules/authors/static/js/authors/author.js [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+  app.directive('authorFields', function() {
+    return {
+      restrict: 'E',
+      template: '<ul class="research-area">' +
+                '<li ng-repeat="field in authorFields">' +
+                '{{ field }}</li></ul>',
+      scope: {
+        fields: '=',
+        // It is an isolated scope. No one can do about it.
```
Comment (jmartinm) [inline inspirehep/modules/authors/static/js/authors/author.js]: No one? Or Nothing?

What is `fields` used for?  I see that below you use `scope.authorFields`

File: inspirehep/modules/authors/static/js/authors/author.js (hunk 10 above)
Comment (jacenkow) [inline inspirehep/modules/authors/static/js/authors/author.js]: `<author-fields record='{{ record|tojson|safe }}' fields="{{ config.ARXIV_TO_INSPIRE_CATEGORY_MAPPING}}"></author-fields>`

File: inspirehep/modules/authors/static/js/authors/author.js (hunk 11 above)
Comment (jacenkow) [inline inspirehep/modules/authors/static/js/authors/author.js]: `<author-fields record='{{ record|tojson|safe }}' fields="{{ config.ARXIV_TO_INSPIRE_CATEGORY_MAPPING}}"></author-fields>`

File: inspirehep/modules/authors/static/js/authors/publications.js (hunk 9 above)
Comment (jmartinm) [inline inspirehep/modules/authors/static/js/authors/publications.js]: One option to consider is like in this example https://github.com/inveniosoftware/invenio-search-js/blob/master/src/invenio-search-js/invenioSearch.module.js#L372 where there is a parent directive and the data is only fetched when the event happens https://github.com/inveniosoftware/invenio-search-js/blob/master/src/invenio-search-js/invenioSearch.module.js#L291

File: inspirehep/modules/authors/static/js/authors/publications.js [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+        });
+      }
+    };
+  });
+
+    app.directive('statistics', function() {
+    return {
+      restrict: 'E',
+      templateUrl: '/static/js/authors/templates/statistics.html',
+      scope: true,
```
Comment (jmartinm) [inline inspirehep/modules/authors/static/js/authors/publications.js]: Also related to the comment before of having a parent directive wrapping all the other ones. It could be a good idea to do it like in this [invenio library](https://github.com/inveniosoftware/invenio-search-js/blob/master/src/invenio-search-js/invenioSearch.module.js) used in search results page. 

There, all the directives that are inside of the parent just have `scope: False` and require the parent - see https://github.com/inveniosoftware/invenio-search-js/blob/master/src/invenio-search-js/invenioSearch.module.js#L558-L564 so that they all share the same controller and scope.

File: inspirehep/modules/authors/static/scss/authors/authors-profile.scss [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Invenio; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+// Colours.
+$background: #EDF3F4;
```
Comment (jmartinm) [inline inspirehep/modules/authors/static/scss/authors/authors-profile.scss]: As we mentioned today, the author profile will be using the general INSPIRE colours from https://github.com/inspirehep/inspire-next/blob/invenio3/inspirehep/modules/theme/static/scss/base/variables.scss

Comment (jmartinm) [inline inspirehep/modules/authors/static/scss/authors/authors-profile.scss]: And you can add the author profile `scss` file to https://github.com/inspirehep/inspire-next/blob/invenio3/inspirehep/modules/theme/static/scss/inspirehep.scss

File: inspirehep/modules/authors/views.py [hunk 14]
```diff
@@ -154,6 +154,79 @@ def get_inspire_url(data):
...
+                    keywords.add(keyword.get('keyword'))
+
+            # Get collaborations.
+            for collaboration in result_source.get('collaboration', []):
+                collaborations.add(collaboration.get('value'))
+
+            if publication:
+                publications.append(publication)
+
+        except KeyError:
```
Comment (jmartinm) [inline inspirehep/modules/authors/views.py]: This seems like a too big `try...except` block. You can just add it around the `title` code.

File: inspirehep/modules/authors/views.py [hunk 15]
```diff
@@ -154,6 +154,79 @@ def get_inspire_url(data):
...
+            # Broken article with no title.
+            pass
+
+    response = {}
+    response['publications'] = publications
+    response['keywords'] = list(keywords)
+    response['collaborations'] = list(collaborations)
+
+    return json.dumps(response)
```
Comment (jmartinm) [inline inspirehep/modules/authors/views.py]: Use `from flask import jsonify` - http://flask.pocoo.org/docs/0.10/api/#flask.json.jsonify

File: inspirehep/modules/theme/ext.py (hunk 7 above)
Comment (jmartinm) [inline inspirehep/modules/theme/ext.py]: Still pending to be removed.

File: inspirehep/modules/theme/templates/inspirehep_theme/format/record/authors/Author_HTML_detailed.html [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+        <div class="col-md-12 summary" id="detailed-header-top">
+          <div class="col-md-2">
+            <img src="{{ url_for('static', filename='images/einstein.svg') }}" width="100%">
+          </div>
+          <div class="col-md-10">
+            <div id="record-title">
+              <author-name></author-name>
+              <author-affiliation id="affiliation"></author-affiliation>
+            </div>
+            <div id="fields">
```
Comment (jmartinm) [inline inspirehep/modules/theme/templates/inspirehep_theme/format/record/authors/Author_HTML_detailed.html]: Add a prefix to all the `ids` to make sure they will not later conflict. Like `author-fields`

File: inspirehep/modules/authors/static/js/authors/author.js (hunk 11 above)
Comment (jmartinm) [inline inspirehep/modules/authors/static/js/authors/author.js]: You can just access directly those parameters in the `link` function with the `attrs` parameter - see https://github.com/inveniosoftware/invenio-search-js/blob/master/src/invenio-search-js/invenioSearch.module.js#L343

File: inspirehep/modules/theme/templates/inspirehep_theme/format/record/authors/Author_HTML_detailed.html (hunk 16 above)
Comment (jacenkow) [inline inspirehep/modules/theme/templates/inspirehep_theme/format/record/authors/Author_HTML_detailed.html]: Done.

File: inspirehep/modules/authors/views.py (hunk 15 above)
Comment (jacenkow) [inline inspirehep/modules/authors/views.py]: Done.

File: inspirehep/modules/authors/static/js/authors/author.js (hunk 10 above)
Comment (jacenkow) [inline inspirehep/modules/authors/static/js/authors/author.js]: Done.

File: inspirehep/modules/authors/static/js/authors/author.js (hunk 11 above)
Comment (jacenkow) [inline inspirehep/modules/authors/static/js/authors/author.js]: Done.

**Comment by jacenkow** [comment]
Blocker / missing data:
- [ ] co-authors
- [ ] citations
- [ ] HEP index
- [ ] DataTable (for listing publications)
- [ ] `recids` of affiliations and collaborations

To do:
- [ ] proper styling of statistics graph
- [x] hookup publication type (journal, conference paper, etc.)
- [ ] refactor stylesheet
- [x] create a parent controller for handling data
- [ ] remove Jinja escaping
- [ ] investigate `build.js`
- [ ] investigate `try / except`s

**Comment by jacenkow** [comment]
screen shot 2016-04-14 at 16 36 44

Almost.

**Comment by jacenkow** [comment]
image

**Comment by jacenkow** [comment]
No I will try to port the publications list to `DataTable`

**Comment by jacenkow** [comment]
@jmartinm :8ball:

**Comment by jacenkow** [comment]
image

**Comment by jacenkow** [comment]
profile

**Comment by jacenkow** [comment]
Blocker / missing data:
- [ ] co-authors
- [ ] HEP index (@kaplun ?)
- [ ] `recids` of affiliations and ~~collaborations~~

To do:
- [ ] hookup citations count from `HEP` publications
- [ ] hookup impact graphs for each publication
- [ ] add personal links (`Orcid`, personal pages, and so on)
- [ ] remove profile picture as well as keywords while website is run on mobile devices
- [ ] DO TESTS AND ADD COMMENTS

**Comment by jacenkow** [comment]
@jmartinm ready for the review 😄

**Comment by kaplun** [comment]
@jacenkow is the above TODO list up-to-date?

**Comment by jacenkow** [comment]
@kaplun yes.

**Comment by kaplun** [comment]
For H-Index I'd suggest to create a simple utils to generate it based on the number of citation of all the list of publication.

See: http://stackoverflow.com/questions/20139640/looking-for-algorithm-to-calculate-h-index-fast

File: inspirehep/config.py [hunk 17]
```diff
@@ -730,162 +731,162 @@ def _(x):
...
-    'q-fin.TR': 'Other',
-    'quant-ph': 'General Physics',
-    'solv-int': 'Math and Math Physics',
-    'stat.AP': 'Other',
-    'stat.CO': 'Other',
-    'stat.ME': 'Other',
-    'stat.ML': 'Other',
-    'stat.OT': 'Other',
-    'stat.TH': 'Other'
+    "alg-geom": "Math and Math Physics",
```
Comment (jmartinm) [inline inspirehep/config.py]: why the change to double quote here? - when up you use single quote https://github.com/inspirehep/inspire-next/pull/990/files#diff-3f822b17443888990c9f70f9f89d66f6R454

Comment (jacenkow) [inline inspirehep/config.py]: So when I parse `author` object I use single quotes in `Jinja` template, but when it's coming to the config I have to use single quotes. This way I use double quotes in both cases.

File: inspirehep/modules/authors/views.py [hunk 18]
```diff
@@ -154,6 +154,96 @@ def get_inspire_url(data):
...
+            },
+            index='records-hep',
+            doc_type='hep'):
+
+        try:
+            result_source = result['_source']
+            publication = {}
+
+            # Get publication title (required).
+            publication['title'] = result_source.get(
```
Comment (jmartinm) [inline inspirehep/modules/authors/views.py]: You can use https://github.com/inspirehep/inspire-next/blob/master/inspirehep/utils/record.py#L30 so that we unify how to get a title everywhere.

File: inspirehep/modules/authors/static/scss/authors/authors-profile.scss [hunk 19]
```diff
@@ -0,0 +1,202 @@
...
+ *
+ * Invenio is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Invenio; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+ */
```
Comment (jmartinm) [inline inspirehep/modules/authors/static/scss/authors/authors-profile.scss]: We can import this file in https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/theme/static/scss/inspirehep.scss and then you would have access to the general site color variables https://github.com/inspirehep/inspire-next/blob/master/inspirehep/modules/theme/static/scss/base/variables.scss (you can add there extra ones)

**Comment by jmartinm** [comment]
Besides the small comments looks good to me overall

File: inspirehep/modules/authors/static/scss/authors/authors-profile.scss (hunk 19 above)
Comment (jacenkow) [inline inspirehep/modules/authors/static/scss/authors/authors-profile.scss]: Done.

File: inspirehep/modules/authors/views.py (hunk 18 above)
Comment (jacenkow) [inline inspirehep/modules/authors/views.py]: Done.

**Comment by jacenkow** [comment]
@jmartinm done. After you hopefully merge it, I will create an issue to truck stuff to do.

**Comment by kaplun** [comment]
@jacenkow this PR needs to be closed and reopened against master branch.

**Comment by jacenkow** [comment]
@kaplun on its way.

**Comment by jacenkow** [comment]
Continued in #1094
